### PR TITLE
Persisted "Show content of concatenated modules" with localStorage

### PR DESCRIPTION
**Pull request by lorenzos**
As discussed in #321.

To my surprise, this works also when `{ analyzerMode: 'static' }`: I didn't know browsers save `localStorage` for local files too.

### CODE REVIEW

**Comment by valscion** [comment]
Thanks! We should probably guard for `localStorage` methods raising errors, as there are plenty of situations where they could fail and the UI should work anyway

**Comment by th0r** [comment]
Yep, we should create a small module that will wrap all the `localStorage` methods with `try {} catch`.

**Comment by lorenzos** [comment]
Done in latest commit. Instead of creating a new module, I used `utils.js`.

File: client/utils.js [hunk 1]
```diff
@@ -17,3 +17,27 @@ export function walkModules(modules, cb) {
 export function elementIsOutside(elem, container) {
   return !(elem === container || container.contains(elem));
 }
+
+export const localStorage = {
```
Comment (th0r) [inline client/utils.js]: Extract it into a separate module please

File: client/utils.js [hunk 2]
```diff
@@ -17,3 +17,27 @@ export function walkModules(modules, cb) {
...
+
+  getItem(k) {
+    try {
+      return window.localStorage.getItem(k);
```
Comment (th0r) [inline client/utils.js]: Let's create a constant that will contain a prefix e.g. `wba` and prepend it to all keys. 
Let's also `JSON.stringify/parse` values so we can store any primitive value, not just string.

This is how I see this function:
```js
getItem(key) {
  try {
    return JSON.parse(window.localStorage.getItem(`${KEY_PREFIX}.${key}`));
  catch (err) {
    return null;
  }
}
```

File: client/utils.js [hunk 3]
```diff
@@ -17,3 +17,27 @@ export function walkModules(modules, cb) {
...
+    } catch (x) {
+      return null;
+    }
+  },
+
+  setItem(k, v) {
```
Comment (th0r) [inline client/utils.js]:
```suggestion
  setItem(key, value) {
```

File: client/utils.js [hunk 4]
```diff
@@ -17,3 +17,27 @@ export function walkModules(modules, cb) {
...
+    try {
+      window.localStorage.setItem(k, v);
```
Comment (th0r) [inline client/utils.js]:
```suggestion
      window.localStorage.setItem(key, JSON.stringify(value));
```

File: client/utils.js [hunk 5]
```diff
@@ -17,3 +17,27 @@ export function walkModules(modules, cb) {
...
+    } catch (x) { /* ignored */ }
+  },
+
+  removeItem(k) {
```
Comment (th0r) [inline client/utils.js]:
```suggestion
  removeItem(key) {
```

File: client/utils.js [hunk 6]
```diff
@@ -17,3 +17,27 @@ export function walkModules(modules, cb) {
...
+    try {
+      window.localStorage.removeItem(k);
```
Comment (th0r) [inline client/utils.js]:
```suggestion
      window.localStorage.removeItem(key);
```

File: client/components/ModulesTreemap.jsx [hunk 7]
```diff
@@ -208,6 +208,11 @@ export default class ModulesTreemap extends Component {
 
   handleConcatenatedModulesContentToggle = flag => {
     store.showConcatenatedModulesContent = flag;
+    if (flag) {
+      localStorage.setItem('showConcatenatedModulesContent', '1');
```
Comment (th0r) [inline client/components/ModulesTreemap.jsx]: Becomes this after implementing suggestions below:
```suggestion
      localStorage.setItem('showConcatenatedModulesContent', true);
```

**Comment by lorenzos** [comment]
Should be ok now @ 9823e7c

**Comment by valscion** [comment]
Looks good to me, what do you think @th0r?

It would be nice to have a changelog entry for this, similar to what we have for other things. This would be an "Improvement" change

**Comment by lorenzos** [comment]
News on this?

**Comment by th0r** [comment]
> It would be nice to have a changelog entry for this, similar to what we have for other things. This would be an "Improvement" change

@lorenzos

**Comment by lorenzos** [comment]
> @lorenzos could you add a note to `CHANGELOG.md` under `UNRELEASED` section please?

Done. Sorry, I didn't get the request was addressed to me back then :)

**Comment by th0r** [comment]
Thanks, merged!

**Comment by valscion** [comment]
This has been released in v3.7.0 ☺️

**Pull request merged**